### DISCUSSION TITLE
Manipulating image annotations

### DISCUSSION THREAD

**Discussion by ankur655** [stephanrauh/ngx-extended-pdf-viewer#2564] [General]
I was trying to do two things to an image annotation that is programatically added :

1. adding an event handler on clicking the annotation:
document.querySelector('div.stampEditor')?.addEventListener('click', (evt) => {console.log("clicked annotation")});

2. removing the "Add Alt text" and "remove" icons from the image annotation:
document.querySelector('div.editToolbar').style.display = "none";

However both these don't seem to work - I spent hours trying to debug this but couldn't get anywhere. Maybe I am missing something very simple - so would appreciate if somebody could point out what's wrong.

**Comment by stephanrauh** [comment]
Did you check if `document.querySelector(...)` returns a result? If it doesn't, maybe there's a timing problem. Add the event listener and the inline style after a short timeout.

**Reply by ankur655** [reply]
yes , it did return the node correctly. however adding the event handler didn't help.

**Reply by stephanrauh** [reply]
You need to unset the pointEvents style of the annotation layer. But then you also need to deactivate drag & drop and the resizers.
```ts
    const stamp = document.querySelector('div.stampEditor');
    stamp?.addEventListener('click', (evt) => {
      console.log("clicked annotation");
      evt.stopPropagation();
    });
    const annotationLayer = stamp?.parentElement as HTMLElement;
    annotationLayer.style.pointerEvents = "unset";
    const toolbar = document.querySelector('div.editToolbar') as HTMLElement;
    toolbar.style.display = "none";
```

**Reply by ankur655** [reply]
Thanks for your prompt response.  I tested it out and just showing the annotation nothing happens - then when I open the inspect tool in Chrome OR change the pdf zoom, the bottom edit functionality vanishes and the click action is recorded.  

Similarly, if inspect tool is already on, nothing happens then turning it off causes the correct set of things to happen.  

So looks like some kind of rerendering is needed for the javascript to kick in?   I have attached two images,  before changing zoom and after changing zoom to show this.  The 88% Zoom is the initial state,  the 100% zoom is the final state (which shows the desired effect)
<img width="603" alt="capture1" src="https://github.com/user-attachments/assets/3d9679e4-63fd-41e1-9f00-854381761a81">
<img width="502" alt="capture2" src="https://github.com/user-attachments/assets/a39e6dac-d297-4220-9b4e-4afbb8226801">

**Reply by stephanrauh** [reply]
I have to admit I kept the debugger open all the time. However, I don't believe you need a re-rendering. Quite the contrary, re-rendering replaces the HTML div representing the annotation with a new one, so I'd expect the click listener to be gone.

Maybe you have to play with the z-index to make it work.

BTW, AFAIK, you can get rid of the resizer with CSS.

**Reply by ankur655** [reply]
Sorry if I was not clear - but what I intended to say was some toggling (either: turning on the inspect tool OR doing a zoom) is needed to make both the click event handler and removing the bottom edit functionality work.  The javascript doesn't seem to work as is without doing the toggling.

**Reply by stephanrauh** [reply]
On my machine, the event listener worked immediately, without having to do anything.

**Reply by ankur655** [reply]
at what point would the div.stampEditor exist?  It seems it doesn't exist onAnnotationLayerRendered event handler for a small file like https://www.orimi.com/pdf-test.pdf  .  The annotation is created in the onPagesLoaded event handler for the first page.  But I guess that still doesn't guarantee that it's already in the DOM tree before the annotatoinLayerRednered is called?   On the other hand for  a very long PDF with 100 pages, the onAnnotationLayerRendered event handler shows the div.stampEditor element exists.   Is there another event handler that can help indicate that the annotation is already done? 

Also once the annotation event handler is active, the image becomes draggable - is there a way to prevent this?

**Reply by ankur655** [reply]
I also tried creating the image annotation in onAnnotationLayerRendered  and then waited for an AnnotationEditorEvent but that didn't seem to happen
